In [1]:
import numpy as np
import pandas as pd

In [11]:
df=pd.read_csv("/content/sample_data/data.csv",usecols=['Sentence','Sentiment'])

In [12]:
df# original dataset

Sentence Sentiment
0     The GeoSolutions technology will leverage Bene...  positive
1     $ESI on lows, down $1.50 to $2.50 BK a real po...  negative
2     For the last quarter of 2010 , Componenta 's n...  positive
3     According to the Finnish-Russian Chamber of Co...   neutral
4     The Swedish buyout firm has sold its remaining...   neutral
...                                                 ...       ...
5837  RISING costs have forced packaging producer Hu...  negative
5838  Nordic Walking was first used as a summer trai...   neutral
5839  According shipping company Viking Line , the E...   neutral
5840  In the building and home improvement trade , s...   neutral
5841  HELSINKI AFX - KCI Konecranes said it has won ...  positive

[5842 rows x 2 columns]

**Prerocessing**



In [13]:
df['cleaned_data']=df['Sentence'].str.replace(r'[^\w\s]','',regex=True) #data cleaning by removing all symbols

In [51]:
df # clean_data dataset

Sentence Sentiment  \
0     The GeoSolutions technology will leverage Bene...  positive   
1     $ESI on lows, down $1.50 to $2.50 BK a real po...  negative   
2     For the last quarter of 2010 , Componenta 's n...  positive   
3     According to the Finnish-Russian Chamber of Co...   neutral   
4     The Swedish buyout firm has sold its remaining...   neutral   
...                                                 ...       ...   
5837  RISING costs have forced packaging producer Hu...  negative   
5838  Nordic Walking was first used as a summer trai...   neutral   
5839  According shipping company Viking Line , the E...   neutral   
5840  In the building and home improvement trade , s...   neutral   
5841  HELSINKI AFX - KCI Konecranes said it has won ...  positive   

                                           cleaned_data  
0     The GeoSolutions technology will leverage Bene...  
1     ESI on lows down 150 to 250 BK a real possibility  
2     For the last quarter of 2010  Componenta s net...  
3     According to the FinnishRussian Chamber of Com...  
4     The Swedish buyout firm has sold its remaining...  
...                                                 ...  
5837  RISING costs have forced packaging producer Hu...  
5838  Nordic Walking was first used as a summer trai...  
5839  According shipping company Viking Line  the EU...  
5840  In the building and home improvement trade  sa...  
5841  HELSINKI AFX  KCI Konecranes said it has won a...  

[5842 rows x 3 columns]

**Removing stop words**

In [55]:
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')

def remove_stop_words(string):
    stop_words = set(stopwords.words('english'))
    words = string.split()
    filtered_words = [word for word in words if word.lower() not in stop_words]
    new_string = ' '.join(filtered_words)
    return new_string

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [58]:
df['new_cleaned_data']=df['cleaned_data'].apply(remove_stop_words)
# applying def function and creating new col.without stopwords

In [59]:
df.head()

Sentence Sentiment  \
0  The GeoSolutions technology will leverage Bene...  positive   
1  $ESI on lows, down $1.50 to $2.50 BK a real po...  negative   
2  For the last quarter of 2010 , Componenta 's n...  positive   
3  According to the Finnish-Russian Chamber of Co...   neutral   
4  The Swedish buyout firm has sold its remaining...   neutral   

                                        cleaned_data  \
0  The GeoSolutions technology will leverage Bene...   
1  ESI on lows down 150 to 250 BK a real possibility   
2  For the last quarter of 2010  Componenta s net...   
3  According to the FinnishRussian Chamber of Com...   
4  The Swedish buyout firm has sold its remaining...   

                                    new_cleaned_data  
0  GeoSolutions technology leverage Benefon GPS s...  
1               ESI lows 150 250 BK real possibility  
2  last quarter 2010 Componenta net sales doubled...  
3  According FinnishRussian Chamber Commerce majo...  
4  Swedish buyout firm sold remaining 224 percent...

In [60]:
df.loc[1]
#showing all the difference in 3 col

Sentence            $ESI on lows, down $1.50 to $2.50 BK a real po...
Sentiment                                                    negative
cleaned_data        ESI on lows down 150 to 250 BK a real possibility
new_cleaned_data                 ESI lows 150 250 BK real possibility
Name: 1, dtype: object

In [15]:
from sklearn.feature_extraction.text import CountVectorizer

In [16]:
vectorizer1=CountVectorizer(binary=True)#checks if words present in vocab is present

In [17]:
vectorizer2=CountVectorizer(binary=False)##checks if how many time vocab word is present in document

In [62]:
x3=vectorizer1.fit_transform(df['new_cleaned_data'])

In [63]:
x4=vectorizer2.fit_transform(df['new_cleaned_data'])

In [20]:
y=df['Sentiment']

In [21]:
from sklearn.model_selection import train_test_split

In [75]:
xtrain4,xtest4,ytrain4,ytest4=train_test_split(x3,y,test_size=0.25,random_state=1)#for new_cleaned_data

In [76]:
xtrain5,xtest5,ytrain5,ytest5=train_test_split(x4,y,test_size=0.25,random_state=1)#for new_cleaned_data

In [73]:
from sklearn.naive_bayes import BernoulliNB,MultinomialNB

In [77]:
model1=BernoulliNB()
model2=MultinomialNB()

In [81]:
model1.fit(xtrain4,ytrain4)

BernoulliNB()

In [82]:
model2.fit(xtrain5,ytrain5)

MultinomialNB()

In [102]:
pred3=model1.predict(xtest4)

In [103]:
pred4=model2.predict(xtest5)

In [34]:
from sklearn.metrics import accuracy_score

In [104]:
accuracy_score(ytest,pred3)   # new_cleaned_data using burnoulli

0.6707734428473648

In [105]:
accuracy_score(ytest,pred4) # new_cleaned_data using multinomial

0.6625598904859685

**TDIF**

In [37]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [38]:
from sklearn.pipeline import make_pipeline

In [42]:
model3=make_pipeline(TfidfVectorizer(),MultinomialNB())

In [106]:
x1=df['new_cleaned_data']

In [97]:
xtrain7,xtest7,ytrain7,ytest7=train_test_split(x1,y,test_size=0.25,random_state=1)

In [98]:
model3.fit(xtrain7,ytrain7)

Pipeline(steps=[('tfidfvectorizer', TfidfVectorizer()),
                ('multinomialnb', MultinomialNB())])

In [99]:
pred4=model3.predict(xtest7)

In [100]:
accuracy_score(ytest1,pred4)# for new_cleaned_data

0.6536618754277892

final result:

1.   accuracy score for TFID:0.6536618754277892
2.   accuracy score for Multinomial: 0.6625598904859685,            
Accuracy score for bernouille: 0.6707734428473648

**Saving The Model**

In [109]:
import joblib
joblib.dump(model1,"bernoullie.pkl")

['bernoullie.pkl']

**Loading into variable**

In [110]:
new_model=joblib.load("bernoullie.pkl")